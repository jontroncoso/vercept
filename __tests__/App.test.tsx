/**
 * @jest-environment jsdom
 */

import "@testing-library/jest-dom";
import { act, render, screen } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import App from "../src/App";

describe("App", () => {
  it("renders without crashing", async () => {
    const { container } = render(<App />);

    // Check for the presence of key elements
    const textBox = screen.getByPlaceholderText("Ask anything...");
    const [addImageButton] = container.querySelectorAll("#dropzone button") as NodeListOf<HTMLButtonElement>;
    expect(textBox).toBeInTheDocument();
    expect(addImageButton).toBeInTheDocument();

    // Mock the createElement to handle file input creation
    let input: HTMLInputElement | null = null;
    const createElementSpy = vi.spyOn(document, "createElement").mockImplementationOnce((tagName: string) => {
      if (tagName === "input") {
        input = document.createElement("input");
        input.type = "file";
        return input;
      }
      return document.createElement(tagName);
    });

    // Simulate clicking the add image button
    addImageButton.click();
    expect(createElementSpy).toBeCalled();
    expect(input).not.toBeNull();

    /* Select image to upload */
    const file = new File(["dummy content"], "test-image.png", { type: "image/png" });
    Object.defineProperty(input, "files", { value: [file] });
    expect(container.querySelectorAll("#dropzone img")).toHaveLength(0);
    act(() => input!.dispatchEvent(new Event("change", { bubbles: true })));

    // Expect the image to be rendered in the document
    const images = container.querySelectorAll("#dropzone img");
    expect(images).toHaveLength(1);
    const dropzoneImage = images[0] as HTMLImageElement;
    expect(dropzoneImage.src).toContain("http://localhost:3000/mock-object-url");

    // Click the X to remove the image
    const removeButton = dropzoneImage.nextSibling as HTMLButtonElement;
    act(() => removeButton.click());
    expect(container.querySelectorAll("#dropzone img")).toHaveLength(0);
  });

  it("Ask a question in the textbox", async () => {
    // const { container } =
    render(<App />);

    const textbox = screen.getByPlaceholderText("Ask anything...");
    expect(textbox).toBeInTheDocument();

    // Simulate user typing a question
    act(() => {
      textbox.textContent = "What is in the image?\n";
      textbox.dispatchEvent(new Event("keydown", { bubbles: true }));
    });

    const chatMessages = await screen.findAllByText("What is in the image?");

    expect(chatMessages).toHaveLength(1);
  });
});
