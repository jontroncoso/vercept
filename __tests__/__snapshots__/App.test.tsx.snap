// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Ask a question with images 1`] = `
<output
  aria-label="Chatbot conversation"
  class="overflow-y-scroll flex-1 shrink-0 px-3 pt-2"
>
  <div
    class="flex flex-col relative justify-end items-end mb-4"
  >
    <div
      class="rounded-2xl text-secondary-foreground px-5 py-3 shadow-primary/50 shadow-md text-sm mb-8 bg-chart-1 text-white rounded-tr-none"
    >
      What is in the image?
    </div>
  </div>
  <div
    class="flex flex-col relative justify-end items-start mb-4"
  >
    <div
      class="rounded-2xl text-secondary-foreground px-5 py-3 shadow-primary/50 shadow-md text-sm mb-8 bg-muted text-primary rounded-tl-none"
    >
      Howdy Howdy Howdy!
      <div
        class="flex items-center gap-4 text-muted -mb-12 pt-4"
      >
        <button
          aria-label="Copy"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-copy h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="14"
              rx="2"
              ry="2"
              width="14"
              x="8"
              y="8"
            />
            <path
              d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
            />
          </svg>
        </button>
        <button
          aria-label="Like"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-thumbs-up h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 10v12"
            />
            <path
              d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2a3.13 3.13 0 0 1 3 3.88Z"
            />
          </svg>
        </button>
        <button
          aria-label="Dislike"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-thumbs-down h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 14V2"
            />
            <path
              d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22a3.13 3.13 0 0 1-3-3.88Z"
            />
          </svg>
        </button>
        <button
          aria-label="Read"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-volume2 lucide-volume-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
            />
            <path
              d="M16 9a5 5 0 0 1 0 6"
            />
            <path
              d="M19.364 18.364a9 9 0 0 0 0-12.728"
            />
          </svg>
        </button>
        <button
          aria-label="Refresh"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-refresh-cw h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
            />
            <path
              d="M21 3v5h-5"
            />
            <path
              d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
            />
            <path
              d="M8 16H3v5"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex flex-col relative justify-end items-end mb-4"
  >
    <img
      alt="Chatted image 1"
      class="rounded-lg max-w-xs"
      src="https://chatbot.tronco.so/upload/test-image.png"
    />
    <div
      class="rounded-2xl text-secondary-foreground px-5 py-3 shadow-primary/50 shadow-md text-sm mb-8 bg-chart-1 text-white rounded-tr-none"
    >
      What is in the image?
    </div>
  </div>
  <div
    class="flex flex-col relative justify-end items-start mb-4"
  >
    <div
      class="rounded-2xl text-secondary-foreground px-5 py-3 shadow-primary/50 shadow-md text-sm mb-8 bg-muted text-primary rounded-tl-none"
    >
      Howdy Howdy Howdy!
      <div
        class="flex items-center gap-4 text-muted -mb-12 pt-4"
      >
        <button
          aria-label="Copy"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-copy h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="14"
              rx="2"
              ry="2"
              width="14"
              x="8"
              y="8"
            />
            <path
              d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
            />
          </svg>
        </button>
        <button
          aria-label="Like"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-thumbs-up h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 10v12"
            />
            <path
              d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2a3.13 3.13 0 0 1 3 3.88Z"
            />
          </svg>
        </button>
        <button
          aria-label="Dislike"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-thumbs-down h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 14V2"
            />
            <path
              d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22a3.13 3.13 0 0 1-3-3.88Z"
            />
          </svg>
        </button>
        <button
          aria-label="Read"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-volume2 lucide-volume-2 h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
            />
            <path
              d="M16 9a5 5 0 0 1 0 6"
            />
            <path
              d="M19.364 18.364a9 9 0 0 0 0-12.728"
            />
          </svg>
        </button>
        <button
          aria-label="Refresh"
          class="p-1 rounded-full"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-refresh-cw h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
            />
            <path
              d="M21 3v5h-5"
            />
            <path
              d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
            />
            <path
              d="M8 16H3v5"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</output>
`;

exports[`App > prevents adding duplicate images 1`] = `
<div>
  <div
    class="h-screen w-screen text-foreground flex flex-col"
  >
    <div
      class="overflow-y-scroll flex-1 shrink-0 px-3 pt-2"
    >
      <div
        class="flex relative justify-end items-end"
      >
        <img
          alt="Stack of books"
          class="rounded-2xl border object-contain max-h-96"
          src="https://images.prismic.io/star-trek-untold/NDQzYTYyYWItMjY2Ny00MGY3LWEzMzItMDNkNjdhOWMyMzg2_chain_of_command_2.jpg?auto=compress,format&rect=0,0,700,526&w=700&h=526"
        />
        <div
          class="absolute bottom-6 right-0"
        >
          <div
            class="rounded-full rounded-tr-none text-secondary-foreground bg-secondary px-5 py-3 shadow-primary/20 shadow-md text-sm"
          >
            How many lights are in this image?
          </div>
        </div>
      </div>
      <div
        class="space-y-3"
      >
        <p
          class="text-lg"
        >
          There are four lights in this image.
        </p>
        <div
          class="flex items-center gap-4 text-muted"
        >
          <button
            aria-label="Copy"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-copy h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="2"
                ry="2"
                width="14"
                x="8"
                y="8"
              />
              <path
                d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
              />
            </svg>
          </button>
          <button
            aria-label="Like"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-thumbs-up h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 10v12"
              />
              <path
                d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2a3.13 3.13 0 0 1 3 3.88Z"
              />
            </svg>
          </button>
          <button
            aria-label="Dislike"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-thumbs-down h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 14V2"
              />
              <path
                d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22a3.13 3.13 0 0 1-3-3.88Z"
              />
            </svg>
          </button>
          <button
            aria-label="Read"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-volume2 lucide-volume-2 h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
              />
              <path
                d="M16 9a5 5 0 0 1 0 6"
              />
              <path
                d="M19.364 18.364a9 9 0 0 0 0-12.728"
              />
            </svg>
          </button>
          <button
            aria-label="Refresh"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-refresh-cw h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
              />
              <path
                d="M21 3v5h-5"
              />
              <path
                d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
              />
              <path
                d="M8 16H3v5"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="h-98"
      />
    </div>
    <div
      class="flex items-center gap-3 px-3 py-2"
      id="dropzone"
    >
      <div
        class="flex-1 flex-col rounded-2xl relative shadow-zinc-500 shadow-sm transition bg-popover border text-foreground focus-within:shadow-md focus-within:bg-accent"
      >
        <div
          class="flex justify-end py-1.5 px-2 gap-2 min-h-10"
        >
          <textarea
            class="text-secondary-foreground p-2 outline-none grow resize-none"
            placeholder="Ask anything..."
            rows="1"
          />
        </div>
        <div
          class="flex items-center gap-2 py-2 px-2 z-30"
        >
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-plus h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
              <path
                d="M12 5v14"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-globe h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <path
                d="M12 2a14.5 14.5 0 0 0 0 20 14.5 14.5 0 0 0 0-20"
              />
              <path
                d="M2 12h20"
              />
            </svg>
            Search
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-telescope h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m10.065 12.493-6.18 1.318a.934.934 0 0 1-1.108-.702l-.537-2.15a1.07 1.07 0 0 1 .691-1.265l13.504-4.44"
              />
              <path
                d="m13.56 11.747 4.332-.924"
              />
              <path
                d="m16 21-3.105-6.21"
              />
              <path
                d="M16.485 5.94a2 2 0 0 1 1.455-2.425l1.09-.272a1 1 0 0 1 1.212.727l1.515 6.06a1 1 0 0 1-.727 1.213l-1.09.272a2 2 0 0 1-2.425-1.455z"
              />
              <path
                d="m6.158 8.633 1.114 4.456"
              />
              <path
                d="m8 21 3.105-6.21"
              />
              <circle
                cx="12"
                cy="13"
                r="2"
              />
            </svg>
            Deep research
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-ellipsis h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground hover:bg-primary/90 size-9 rounded-full h-12 w-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-mic h-5 w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 19v3"
          />
          <path
            d="M19 10v2a7 7 0 0 1-14 0v-2"
          />
          <rect
            height="13"
            rx="3"
            width="6"
            x="9"
            y="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`App > renders without crashing 1`] = `
<div>
  <div
    class="h-screen w-screen text-foreground flex flex-col"
  >
    <div
      class="overflow-y-scroll flex-1 shrink-0 px-3 pt-2"
    >
      <div
        class="flex relative justify-end items-end"
      >
        <img
          alt="Stack of books"
          class="rounded-2xl border object-contain max-h-96"
          src="https://images.prismic.io/star-trek-untold/NDQzYTYyYWItMjY2Ny00MGY3LWEzMzItMDNkNjdhOWMyMzg2_chain_of_command_2.jpg?auto=compress,format&rect=0,0,700,526&w=700&h=526"
        />
        <div
          class="absolute bottom-6 right-0"
        >
          <div
            class="rounded-full rounded-tr-none text-secondary-foreground bg-secondary px-5 py-3 shadow-primary/20 shadow-md text-sm"
          >
            How many lights are in this image?
          </div>
        </div>
      </div>
      <div
        class="space-y-3"
      >
        <p
          class="text-lg"
        >
          There are four lights in this image.
        </p>
        <div
          class="flex items-center gap-4 text-muted"
        >
          <button
            aria-label="Copy"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-copy h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="2"
                ry="2"
                width="14"
                x="8"
                y="8"
              />
              <path
                d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
              />
            </svg>
          </button>
          <button
            aria-label="Like"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-thumbs-up h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 10v12"
              />
              <path
                d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2a3.13 3.13 0 0 1 3 3.88Z"
              />
            </svg>
          </button>
          <button
            aria-label="Dislike"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-thumbs-down h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 14V2"
              />
              <path
                d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22a3.13 3.13 0 0 1-3-3.88Z"
              />
            </svg>
          </button>
          <button
            aria-label="Read"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-volume2 lucide-volume-2 h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
              />
              <path
                d="M16 9a5 5 0 0 1 0 6"
              />
              <path
                d="M19.364 18.364a9 9 0 0 0 0-12.728"
              />
            </svg>
          </button>
          <button
            aria-label="Refresh"
            class="p-1 rounded-full"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-refresh-cw h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
              />
              <path
                d="M21 3v5h-5"
              />
              <path
                d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
              />
              <path
                d="M8 16H3v5"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="h-98"
      />
    </div>
    <div
      class="flex items-center gap-3 px-3 py-2"
      id="dropzone"
    >
      <div
        class="flex-1 flex-col rounded-2xl relative shadow-zinc-500 shadow-sm transition bg-popover border text-foreground focus-within:shadow-md focus-within:bg-accent"
      >
        <div
          class="flex justify-end py-1.5 px-2 gap-2 min-h-10"
        >
          <textarea
            class="text-secondary-foreground p-2 outline-none grow resize-none"
            placeholder="Ask anything..."
            rows="1"
          />
        </div>
        <div
          class="flex items-center gap-2 py-2 px-2 z-30"
        >
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-plus h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
              <path
                d="M12 5v14"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-globe h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <path
                d="M12 2a14.5 14.5 0 0 0 0 20 14.5 14.5 0 0 0 0-20"
              />
              <path
                d="M2 12h20"
              />
            </svg>
            Search
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-telescope h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m10.065 12.493-6.18 1.318a.934.934 0 0 1-1.108-.702l-.537-2.15a1.07 1.07 0 0 1 .691-1.265l13.504-4.44"
              />
              <path
                d="m13.56 11.747 4.332-.924"
              />
              <path
                d="m16 21-3.105-6.21"
              />
              <path
                d="M16.485 5.94a2 2 0 0 1 1.455-2.425l1.09-.272a1 1 0 0 1 1.212.727l1.515 6.06a1 1 0 0 1-.727 1.213l-1.09.272a2 2 0 0 1-2.425-1.455z"
              />
              <path
                d="m6.158 8.633 1.114 4.456"
              />
              <path
                d="m8 21 3.105-6.21"
              />
              <circle
                cx="12"
                cy="13"
                r="2"
              />
            </svg>
            Deep research
          </button>
          <button
            class="inline-flex items-center gap-2 rounded-full border px-2 py-1 text-sm text-muted min-h-6 min-w-6 leading-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-ellipsis h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground hover:bg-primary/90 size-9 rounded-full h-12 w-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-mic h-5 w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 19v3"
          />
          <path
            d="M19 10v2a7 7 0 0 1-14 0v-2"
          />
          <rect
            height="13"
            rx="3"
            width="6"
            x="9"
            y="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
